### ISSUE TITLE
Audio device failed to suspend with HDA support enabled

### ISSUE THREAD

**Issue by ZhendanYang** [thesofproject/linux#140]
Summary:
Check runtime_status, it always shows “active” whether audio is working or not.
on APL:
cat /sys/bus/pci/devices/0000\:00\:0e.0/power/runtime_status
on CNL:
cat /sys/bus/pci/devices/0000\:00\:1f.3/power/runtime_status

Test env:
kernel: sof-dev: ae5f868
sof: stable 1.2: f7ea2e6
soft: d77beaf

I test with kernel commit(64cf3e3), and the runtime_status works well.

**Comment by ranj063** [comment]
@ZhendanYang could you please check if https://github.com/thesofproject/linux/pull/141 fixes this issue?

**Comment by ZhendanYang** [comment]
@ranj063 
I checked with lasted code base(commit c3ba91f782707bd3e51c46f37300bd7b3cfe8333) which contains your PR. This issue still exist on both APL and CNL.

**Comment by ranj063** [comment]
@ZhendanYang which board are you using for testing on APL? Up squared?

**Comment by ZhendanYang** [comment]
@ranj063 Yes.

**Comment by bardliao** [comment]
I can't reproduce this issue on up2 with kernel: sof-dev:  c3ba91f and sof: master 40f9749. It shows "suspended" when I cat runtime_status after boot.

**Comment by mengdonglin** [comment]
@ZhendanYang could you help check if this is still reproducible on UP2?

**Comment by ZhendanYang** [comment]
It still can reproduce on my APL(UP2) and CNL with kernel sof-dev: c3ba91f and sof master 40f9749 or stable-1.2 0875f0c .
It shows "active" when I cat runtime_status after boot.

**Comment by ranj063** [comment]
@ZhendanYang are you using sof-apl-nocodec or sof-apl-pcm512x?

**Comment by ranj063** [comment]
@ZhendanYang I have root caused this to the problem of having the CONFIG_SND_SOF_HDA enabled in the kernel config.
But with my commit. https://github.com/thesofproject/linux/commit/c3ba91f782707bd3e51c46f37300bd7b3cfe8333, compilation will fail without this in the config. I'll update the issue once I have a solution.

**Comment by bardliao** [comment]
@ZhendanYang I can reproduce it now.

**Comment by ZhendanYang** [comment]
@bardliao @ranj063 
I've tried pcm512x codec, too. It also shows "active" when I cat runtime_status after boot.
And here is my kernel .config file : 
[config.txt](https://github.com/thesofproject/linux/files/2395125/config.txt)

**Comment by ranj063** [comment]
@ZhendanYang could you please make sure CONFIG_SND_SOC_SOF_HDA is disabled in your kernel config and try this commit https://github.com/thesofproject/linux/pull/148

**Comment by ZhendanYang** [comment]
@ranj063 
I disabled CONFIG_SND_SOC_SOF_HDA and try commit #148 , and runtime_status worked well.
@keyonjie 
I'll test again after you sent a new PR.

**Comment by mengdonglin** [comment]
Reopen this issue because we need to assure compatibility here: no matter CONFIG_SND_SOC_SOF_HDA is enabled or not, UP2 topology and machine driver does not include any data path for HDA, so we must make sure I2S and no codec are not broken.

@keyonjie, would you please share you new PR# ?

**Comment by keyonjie** [comment]
@ranj063 mentioned to me that there is a commit from list which may be related with this, maybe we need add it also.
FROMLIST: SoC: Intel: Power down links before tuning off display audio power

**Comment by mengdonglin** [comment]
@keyonjie Could you help to check if this patch will help? If yes, how to make it in sof-dev?

I think @ZhendanYang can verify it after patch is in sof-dev. And in addition, HDMI audio also works on UP2.

**Comment by keyonjie** [comment]
sure, it works on my leafhill, should I send it to Zhendan or Qingliang to give a pre-merge try on up2? @mengdonglin

**Comment by ranj063** [comment]
@ZhendanYang @keyonjie I just checked with the SOF/SOFT master branch and kernel topic/sof-dev branch and runtime PM seems to work fine on my up2. Could you please check as well?

**Comment by ZhendanYang** [comment]
@ranj063 I can't test runtime PM since there is "DSP panic" on up2 after system boot with latest code base. However, on CNL this issue still exists. Could you please provide me your test commit so I can check it for you?

**Comment by keyonjie** [comment]
@ZhendanYang are you running SOF master or glk-003-drop-stable? looks glk-003-drop-stable is better.

**Comment by ZhendanYang** [comment]
@keyonjie glk-003-drop-stable has errors like "ipc error", "topology load fail". I tried runtime_status and it still showed "active".

**Comment by ranj063** [comment]
@ZhendanYang I didnt see any DSP panic's but I was using the eq topology 
sof-apl-eq-pcm512x.tplg

**Comment by ZhendanYang** [comment]
@ranj063 Yes. I tried sof-apl-eq-pcm512x.tplg on up2+hifiberry, runtime_status still showed "active" after system booted.

**Comment by ZhendanYang** [comment]
This issue can't reproduce anymore. Should be fixed by #186 .

**Comment by bardliao** [comment]
@ZhendanYang Yes, the issue is fixed. 
**But root cause is not #186 . The issue is fixed by #165 .**
 The root cause is "HDMI HDA Codec" driver is probed but codec is not probed.
"HDMI HDA Codec" is a child device of "sof-audio-pci". And, "sof-audio-pci" will be active if any of its child is active.
`pm_runtime_enable` is called in `hdmi_codec_probe` 
hdmi_codec_probe will called when sound card is probed.

**Comment by keyonjie** [comment]
so looks like we will still see this issue if we use tplg file without hdmi pipelines?

**Comment by bardliao** [comment]
Yes, we will still see this issue if sound card is not probed.

**Comment by mengdonglin** [comment]
@bardliao @keyonjie  Thank you! It seems this issue is covered by HDMI enabling on UP2, but not fixed.
Let's wait for nocodec boot failure is fixed on UP2 (https://github.com/thesofproject/sof/issues/443) and then double check this issue and re-open this if need.

**Comment by bardliao** [comment]
@mengdonglin Should we disable HDMI support from menuconfig if there is no HDMI in nocodec mode?

**Comment by markyang** [comment]
This issue can't be reproduced on APL.
**Test env:**
sof master: 43da1696
sof tool: 4ac9248
kernel debug/v4.19-backport: 925637e3
tplg: sof-apl-pcm512x.tplg-4ac9248

**Comment by mengdonglin** [comment]
@markyang Could you help to reproduce this issue on both APL (w/ and w/o codec) and CNL? If neither of the devices can reproduce the issue, please close this issue. Thanks!

**Comment by markyang** [comment]
This issue can't be reproduced on APL with codec, CNL & APL w/o codec are blocked by [thesofproject/linux#592](https://github.com/thesofproject/sof/issues/592)

**Comment by markyang** [comment]
@stevyan

**Comment by markyang** [comment]
Created a new bug(thesofproject/linux/issues/512) to follow this issue on APL-nocodec
